pkg/core/resources/store/transactions.go

```diff
@@ -0,0 +1,73 @@
+package store
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+)
+
+type txCtx struct{}
+
+func CtxWithTx(ctx context.Context, tx Transaction) context.Context {
+	return context.WithValue(ctx, txCtx{}, tx)
+}
+
+func TxFromCtx(ctx context.Context) (Transaction, bool) {
+	if value, ok := ctx.Value(txCtx{}).(Transaction); ok {
+		return value, true
+	}
+	return nil, false
+}
+
+type Transaction interface {
+	Commit(ctx context.Context) error
+	Rollback(ctx context.Context) error
+}
+
+// Transactions is a producer of transactions if a resource store support transactions.
+// Transactions are never required for consistency in Kuma, because there are ResourceStores that does not support transactions.
+// However, in a couple of cases executing queries in transaction can improve the performance.
+//
+// In case of Postgres, you may set hooks when retrieve and release connections for the connection pool.
+// In this case, if you create multiple resources, you need to acquire connection and execute hooks for each create.
+// If you create transaction for it, you execute hooks once and you hold the connection.
+//
+// Transaction is propagated implicitly through Context.
+type Transactions interface {
+	Begin(ctx context.Context) (Transaction, error)
```

Comment:
Do we even need transactions? I think we can introduce some kind of "batching" purely on Kuma side – acquire connection in `Begin` and execute all the following operation using the same connection. DB doesn't need to know about these "batches" at all

Reply:
Not sure if I get your proposal here. Can you elaborate? What would it look like?

Reply:
In `Begin()` method, instead of:
```go
return r.pool.Begin(ctx)
```
we can do:
```go
return r.pool.Acquire(ctx) 
```

After that, method `InTx()` puts connection into ctx (instead of putting transaction into ctx).

So later `pgxResourceStore` checks if there is a connection in the context and uses this connection to call `Exec`.

That way a number of store operations are executed on the same connection, but Postgres doesn't know about it.

Reply:
According to multiple sources, doing this in one transaction is more optimal on postgres CPU. 

https://www.postgresql.org/docs/current/populate.html
> 14.4.1. Disable Autocommit 
When using multiple INSERTs, turn off autocommit and just do one commit at the end. (In plain SQL, this means issuing BEGIN at the start and COMMIT at the end. Some client libraries might do this behind your back, in which case you need to make sure the library does it when you want it done.) If you allow each insertion to be committed separately, PostgreSQL is doing a lot of work for each row that is added. An additional benefit of doing all insertions in one transaction is that if the insertion of one row were to fail then the insertion of all rows inserted up to that point would be rolled back, so you won't be stuck with partially loaded data.